Source/DafnyCore/Compilers/SinglePassCompiler.cs

```diff
@@ -2650,10 +2650,11 @@ protected class NullClassWriter : IClassWriter {
         // finally, the jump back to the head of the function
         EmitJumpToTailCallStart(wr);
 
-      } else if (expr is BinaryExpr bin && bin.AccumulatesForTailRecursion != BinaryExpr.AccumulationOperand.None) {
-        Contract.Assert(accumulatorVar != null);
-        Contract.Assert(enclosingFunction != null);
-        Contract.Assert(enclosingFunction.IsAccumulatorTailRecursive);
+      } else if (expr is BinaryExpr bin
+                 && bin.AccumulatesForTailRecursion != BinaryExpr.AccumulationOperand.None
+                 && accumulatorVar != null
+                 && enclosingFunction != null
+                 && enclosingFunction.IsAccumulatorTailRecursive) {
```

Comment:
It's surprising to me that the whole invariant needs to be converted to a check.  Are these really all independent parameters?